shapely/_geometry_helpers.pyx

```diff
@@ -392,6 +440,227 @@ def collections_1d(object geometries, object indices, int geometry_type = 7, obj
+        GEOSGeometry *ring = NULL
+        GEOSGeometry *geom = NULL
+
+    n_geoms = offsets2.shape[0] - 1
+
+    # A temporary array for the geometries that will be given to CreatePolygon.
+    # For simplicity, we use n_geoms instead of calculating
+    # the max needed size (trading performance for a bit more memory usage)
+    temp_rings = np.empty(shape=(n_geoms, ), dtype=np.intp)
+    cdef np.intp_t[:] temp_rings_view = temp_rings
+    # A temporary array for resulting geometries
+    temp_geoms = np.empty(shape=(n_geoms, ), dtype=np.intp)
+    cdef np.intp_t[:] temp_geoms_view = temp_geoms
+
+    # The final target array
+    result = np.empty(shape=(n_geoms, ), dtype=object)
+    cdef object[:] result_view = result
+
+    cdef unsigned int dims = coordinates.shape[1]
+    if dims not in {2, 3}:
+        raise ValueError("coordinates should N by 2 or N by 3.")
+
+    cdef int ring_type = 2
+    cdef int geometry_type = 6  # MultiPolygon
+    cdef char is_ring = 1
+    cdef int handle_nan = 0
+
+    with get_geos_handle() as geos_handle:
+        with nogil:
+            # iterating through the Polygons
+            for i in range(n_geoms):
+
+                # each part (polygon) can consist of multiple rings
+                # (exterior ring + potentially interior rings(s))
+                i1 = offsets2[i]
+                i2 = offsets2[i + 1]
+
+                # iterating through the rings
+                rings_idx = 0
+                for k in range(i1, i2):
+
+                    # each ring consists of certain number of coords
+                    k1 = offsets1[k]
+                    k2 = offsets1[k + 1]
+                    n_coords = k2 - k1
+                    errstate = _create_simple_geometry(
+                        geos_handle, coordinates, k1, n_coords, dims, ring_type,
+                        is_ring, handle_nan, &ring
+                    )
+                    if errstate != PGERR_SUCCESS:
+                        _deallocate_arr(geos_handle, temp_rings_view, rings_idx)
+                        _deallocate_arr(geos_handle, temp_geoms_view, i - 1)
+                        with gil:
+                            return _create_simple_geometry_raise_error(errstate)
+
+                    temp_rings_view[rings_idx] = <np.intp_t>ring
+                    rings_idx += 1
+
+                if rings_idx > 0:
+                    geom = GEOSGeom_createPolygon_r(
+                        geos_handle,
+                        <GEOSGeometry*> temp_rings_view[0],
+                        <GEOSGeometry**> &temp_rings_view[1 if rings_idx > 1 else 0],
+                        rings_idx - 1
+                    )
+                else:
+                    geom = GEOSGeom_createEmptyPolygon_r(geos_handle)
+                if geom == NULL:
+                    _deallocate_arr(geos_handle, temp_rings_view, rings_idx - 1)
+                    _deallocate_arr(geos_handle, temp_geoms_view, i - 1)
+                    with gil:
```

Comment:
Yes, it is required (cython will otherwise give a compilation error). The reason for this is that when returning, it does error checking (check if python has set an error to raise), and that requires the GIL as well (you could think that cython does that automatically on such a bare return, but so that seems not the case)